src/map/pc.cpp

```diff
@@ -12058,253 +12169,544 @@ static unsigned int pc_calc_basesp(uint16 level, uint16 class_) {
+
+		if (!exists) {
+			job->aspd_base.resize(MAX_WEAPON_TYPE + shield);
+			std::fill(job->aspd_base.begin(), job->aspd_base.end(), 2000);
+		}
+
+		for (const auto &aspdit : aspdNode) {
+			std::string weapon = aspdit.first.as<std::string>(), weapon_constant = "W_" + weapon;
+			int64 constant;
+
+			if (!script_get_constant(weapon_constant.c_str(), &constant) || constant < W_FIST || constant > MAX_WEAPON_TYPE + shield) {
+				this->invalidWarning(aspdNode["BaseASPD"], "Invalid weapon type %s specified for %s, skipping.\n", weapon.c_str(), job_name.c_str());
+				continue;
+			}
+
+			int16 aspd;
+
+			if (!this->asInt16(aspdNode, weapon.c_str(), aspd))
+				return 0;
+
+			job->aspd_base[static_cast<int16>(constant)] = aspd;
+		}
 	}
-	idx = pc_class2idx(job_id);
-
-	job_info[idx].max_level[type] = maxlvl;
-	for(i=0; i<maxlvl; i++){
-		t_exp exp = strtoull( fields[3 + i], nullptr, 10 );
-
-		if( exp == 0 ){
-			ShowWarning( "pc_readdb_job_exp: No value defined for level %d in line %d. Defaulting to MAX_EXP...\n", i + 1, current );
-			exp = MAX_EXP;
-		}else if( exp > MAX_EXP ){
-			ShowWarning( "pc_readdb_job_exp: Value %" PRIu64 " is too high, capping to %" PRIu64 "...\n", exp, MAX_EXP );
-			exp = MAX_EXP;
-		}
-
-		job_info[idx].exp_table[type][i] = exp;
-	}
-	//Reverse check in case the array has a bunch of trailing zeros... [Skotlex]
-	//The reasoning behind the -2 is this... if the max level is 5, then the array
-	//should look like this:
-	//0: x, 1: x, 2: x: 3: x 4: 0 <- last valid value is at 3.
-	while ((ui = job_info[idx].max_level[type]) >= 2 && job_info[idx].exp_table[type][ui-2] <= 0)
-		job_info[idx].max_level[type]--;
-	if (job_info[idx].max_level[type] < maxlvl) {
-		ShowWarning("pc_readdb_job_exp: Specified max %u for job %d, but that job's exp table only goes up to level %u.\n", maxlvl, job_id, job_info[idx].max_level[type]);
-		ShowInfo("Filling the missing values with the last exp entry.\n");
-		//Fill the requested values with the last entry.
-		ui = (job_info[idx].max_level[type] <= 2? 0: job_info[idx].max_level[type]-2);
-		for (; ui+2 < maxlvl; ui++)
-			job_info[idx].exp_table[type][ui] = job_info[idx].exp_table[type][ui-1];
-		job_info[idx].max_level[type] = maxlvl;
-	}
-//	ShowInfo("%s - Class %d: %d\n", type?"Job":"Base", job_id, job_info[idx].max_level[type]);
-	for (i = 1; i < job_count; i++) {
-		job_id = jobs[i];
-		if (!pcdb_checkid(job_id)) {
-			ShowError("pc_readdb_job_exp: Invalid job ID %d.\n", job_id);
-			continue;
+
+	if (this->nodeExists(node, "BonusStats")) {
+		const YAML::Node &bonusNode = node["BonusStats"];
+		job->job_bonus.resize(MAX_LEVEL);
+
+		for (const YAML::Node &levelNode : bonusNode) {
+			uint16 level;
+
+			if (!this->asUInt16(levelNode, "Level", level))
+				return 0;
+
+			if (level > MAX_LEVEL) {
+				this->invalidWarning(levelNode["Level"], "Level must be between 1~MAX_LEVEL for %s.\n", job_name.c_str());
+				return 0;
+			}
+
+			if (this->nodeExists(levelNode, "Str")) {
```

Comment:
Would be nice to make this below a little dynamic, since all of them are doing the same thing.
Maybe sub function and handing over the node name and the param constant?